bin/runtime-common/src/messages.rs

```diff
@@ -1183,4 +1194,56 @@ mod tests {
 			.collect()),
 		);
 	}
+
+	#[test]
+	fn verify_messages_proof_with_parser_does_not_panic_if_messages_count_mismatches() {
+		assert_eq!(
+			target::verify_messages_proof_with_parser::<OnThisChainBridge, _, _>(
+				(
+					Default::default(),
+					StorageProof::new(vec![]),
+					Default::default(),
+					0,
+					u64::MAX
+				),
+				0,
+				|_, _| Ok(TestMessageProofParser {
+					failing: false,
+					messages: 0..=u64::MAX,
+					outbound_lane_data: Some(OutboundLaneData {
+						oldest_unpruned_nonce: 1,
+						latest_received_nonce: 1,
+						latest_generated_nonce: 1,
+					}),
+				}),
+			),
+			Err(target::MessageProofError::MessagesCountMismatch),
+		);
+	}
+
+	#[test]
+	#[should_panic]
+	fn verify_messages_proof_with_parser_panic_if_too_many_messages_declared() {
```

Comment:
I'm a little confused here. If by declaring `u64::MAX` messages a user is able to trigger the same Vec allocation as before, doesn't the problem still exist?

Reply:
User won't be able to declare `u64::MAX`, because this code is callable via extrinsic and `messages_count` is a part of weight formula. See description for details

Reply:
So now the unbounded allocation is gone and the condition which would have triggered the unbounded allocation is not user exposed.